adds scipy.stats.mode-esque function

big-c-note opened the pull request:
Solution for (#8978)

This PR seemed stale (#9141), and there were some issues with the shape of the result, which is why the tests were failing. It was easier to start a fresh branch.

I'm new to Jax. I appreciate any feedback you're willing to give.

Super cool project

--
Note: Apologies in advance for the many branches. 

Google's cla scan is not seeing my commit email as being related to the cla I signed. Hopefully rewriting and starting a new branch will trigger it.

jakevdp committed:
jnp.diagonal: raise explicit error if ndim < 2

jakevdp committed:
[sparse] remove deprecated bcoo_add_batch_dim utility

sharadmv committed:
Add loop invariant residual fixpoint test

gnecula committed:
[dynamic_shapes] Fix the lowering of shapes in x64 mode.

The mhlo.reshape op has the constraint that the operand and the result
must have the same element type. Cast the dimension size to int32 to
meet this constraint even in 64-bit mode.

sharadmv committed:
Rollback of HCB GPU custom call due to internal failures

PiperOrigin-RevId: 460079787

jekbradbury committed:
Fix RST formatting in random.py docstring

hawkinsp committed:
Increase bazel sharding of GPU tests.

Reduces the maximum time for some test shards to avoid flaky timeouts.

jakevdp committed:
Improve documentation of jnp.empty() and jnp.empty_like()

yashk2810 committed:
Make `jaxpr_has_pmap` work for other primitives too

PiperOrigin-RevId: 460286042

jakevdp committed:
CI: upgrade doc requirements

jakevdp committed:
Fix execution issues in type_promotion doc

d0k committed:
Integrate LLVM at llvm/llvm-project@71c9757474c3

Updates LLVM usage to match
[71c9757474c3](https://github.com/llvm/llvm-project/commit/71c9757474c3)

PiperOrigin-RevId: 460299215

jakevdp committed:
jnp.ndarray.clip: remove deprecated arguments

yashk2810 committed:
* Check if the device assignment is the same across input and output shardings.
* Allow mixed inputs only if the sharding matches with what is specified in in_axis_resources.

PiperOrigin-RevId: 460326054

pschuh committed:
Use weakref to avoid cache leaks in fori_loop.

This is in response to: https://github.com/google/jax/issues/11448.

maxwillzq committed:
Fix limitations-of-call_tf github link typo.

call_tf misspell as "call-tf" on multiple places.

PiperOrigin-RevId: 460335218

Commit:
Define PyTreeDef on pytree module rather than its parent

PiperOrigin-RevId: 460342835

marcvanzee committed:
[jax2tf] Use tf.nest.flatten instead of tf.keras.backend.flatten.

PiperOrigin-RevId: 460395701

gnecula committed:
[dynamic-shapes] Ensure that the axis_size_env is passed to sub lowering contexts

Commit:
Internal change

PiperOrigin-RevId: 460434859

gnecula committed:
[jax2tf] Fixes for handling of convolutions with shape_polymorphism and enable_xla=False

Issue: #11402

Due to a typo we were running no tests for convolutions with shape
polymorphism and enable_xla=False.

Added a few more tests from #11402 (Thanks @sdenton4).

The main issue was that in presence of shape polymorphism we cannot
just use `x.shape` for a TF value `x` because it will contain `None`
in the place of unknown dimensions. We must use instead the JAX
abstract values.

This does not fix all issues reported in #11402, there is still the
computation of padding or padding="SAME". Commented out the
corresponding test.

jakevdp committed:
BUG: fix jvp rule for lax.rem

atondwal committed:
[mhlo] SubOp -> SubtractOp

Aligns the op class name with the mnemonic

PiperOrigin-RevId: 460560837

jakevdp committed:
add upstream-nightly CI job

big-c-note committed:
adds scipy.stats.mode-esque function

big-c-note commented:
So there is some bug in the cla scan that looks for the first email associated with the cla or takes some time to find the second. Editing the author on the commit alone did not fix the failed cla check, but starting a new branch with the corrected email did.

yashk2810 committed:
Fix the regression introduced by recent changes. Optimize the non-array path in pjit.

PiperOrigin-RevId: 460656090

gnecula committed:
Address review comments

jakevdp commented:
Sorry for the CLA issue – the problem is that *every* commit in the PR needs to be from a known author with a signed CLA, so adding a second commit with a different author email is not sufficient. Glad you finally figured it out! For future reference, if you need to "start from scratch" on a branch, you can do so by force-pushing to the existing branch rather than by opening a new PR.

Thanks for taking a look at this! I'll review shortly 😀

jakevdp commented on jax/_src/scipy/stats/_core.py:
```diff
@@ -0,0 +1,61 @@
+# Copyright 2018 Google LLC
```
2022

jakevdp commented on jax/_src/scipy/stats/_core.py:
```diff
@@ -0,0 +1,61 @@
+# limitations under the License.
+from collections import namedtuple
+from functools import partial
+from typing import Optional, List, Tuple, Iterable, Any
```
nit: alphabetize imports

jakevdp commented on jax/_src/scipy/stats/_core.py:
```diff
@@ -0,0 +1,61 @@
+
+
+@partial(jit, static_argnums=(1,))
+def _mode(x: Iterable[Any], axis: Optional[int]) -> ModeResult:
```
I don't think `Iterable[Any]` is the correct annotation here - for example, you use`x.size` attribute below, which is not in general present on an iterable.

Should it be `x: jnp.ndarray`?

big-c-note replied:
Eg; ```from numpy.typing import ArrayLike```
https://numpy.org/devdocs/reference/typing.html

You're right, it's not quite right. It's probably semantics, but I felt typing as `jnp.ndarray` limiting given this function should handle anything cleared by the `_check_arraylike` func.

I'll go with `jnp.ndarray` for now, but it may be a point to define exactly in typing form what `_check_arraylike` allows. Not sure if there's enough usecase for a typing module for Jax.

big-c-note replied:
And it will definitely be `jnp.ndarray` based on one of your next points, for the hidden functions.

jakevdp commented on jax/_src/scipy/stats/_core.py:
```diff
@@ -0,0 +1,61 @@
+
+@_wraps(scipy.stats.mode)
+def mode(x: Iterable[Any], axis: int = 0, nan_policy: str = "propagate") -> ModeResult:
+  _check_arraylike("mode", x)
```
After `_check_arraylike`, please call `x = jnp.asarray(x)` to ensure that `x` is converted to a `DeviceArray`.

Also, there is a `canonicalize_axis` utility used elsewhere that I think would be relevant here. Regarding handling of `axis=None`, what would you think about doing it this way, so that `axis` can always be an integer in the helper? https://github.com/google/jax/blob/f0b139a59a749f6b3a4776acc50c7a2fa57e9a9e/jax/_src/numpy/reductions.py#L565-L567

A benefit of that I think is that you'd end up calling the `_mode_helper` code in only a single location, so it may lead to a cleaner implementation.

big-c-note replied:
I'll trust your thoughts here. My only concern would be the performance implications.

An exaggerated example being calling `mode` successively in a loop, always with `axis=None` with a large `x`.

Unfortunately, I have a Radeon VII on this machine, so I don't think I could profile that appropriately right now.

jakevdp commented on jax/_src/scipy/stats/_core.py:
```diff
@@ -0,0 +1,61 @@
+def mode(x: Iterable[Any], axis: int = 0, nan_policy: str = "propagate") -> ModeResult:
+  _check_arraylike("mode", x)
+  if nan_policy not in {"propagate", "omit", "raise"}:
+    raise ValueError("Illegal nan_policy value")
```
```suggestion
    raise ValueError(f"Illegal nan_policy value {nan_policy!r}; expected one of {'propoagate', 'omit', 'raise'}")
```

jakevdp commented on jax/_src/scipy/stats/_core.py:
```diff
@@ -0,0 +1,61 @@
+  if nan_policy not in {"propagate", "omit", "raise"}:
+    raise ValueError("Illegal nan_policy value")
+  if nan_policy == "omit":
+    raise NotImplementedError()
```
Please add a message for the user - also perhaps a TODO to implement this? I think it could be done in a JIT-compatible way if we're careful.

big-c-note replied:
For sure, I can tackle that next. Adding TODO

jakevdp commented on jax/_src/scipy/stats/_core.py:
```diff
@@ -0,0 +1,61 @@
+    raise NotImplementedError()
+  if nan_policy == "raise":
+    contains_nan = jnp.isnan(jnp.sum(x))
+    if contains_nan:
```
This will fail with an obscure error under JIT. We should raise a more informative error instead. We handle a similar issue here: https://github.com/google/jax/blob/f0b139a59a749f6b3a4776acc50c7a2fa57e9a9e/jax/_src/numpy/lax_numpy.py#L795-L798

big-c-note replied:
My understanding is that this conditional branch depends on the state of `x` (ie; that it contains nans or not), which defeats the purpose of JIT compiling the `mode` function. 

So, if I am understanding you correctly, this would fail whether or not there were nans present - it's failing because in order to have a tracer object, we can't know what the contents are, only the shape.

jakevdp commented on jax/_src/scipy/stats/_core.py:
```diff
@@ -0,0 +1,61 @@
+    contains_nan = jnp.isnan(jnp.sum(x))
+    if contains_nan:
+      raise ValueError("x contains nans.")
+  return _mode(x=x, axis=axis)
```
I think we could inline the contents of the helper function and jit-compile the `mode` function itself. The main reason that you see the helper pattern throughout JAX's codebase is because for a long time, JIT didn't properly support keyword arguments, so we couldn't use `jit` as a wrapper of top-level API functions that accept keywords. These days that's not a worry.

jakevdp commented on tests/scipy_stats_test.py:
```diff
@@ -889,5 +891,110 @@ def evaluate_kde(kde, x):
+    args_maker = lambda: [jnp.round(rng(x_shape, x_dtype)), axis, nan_policy]
+    # Check on nanpolicy in {'omit'}
+    if nan_policy == "omit":
+      with pytest.raises(NotImplementedError):
```
Here and below, please use `with self.assertRaises` (or better `self.assertRaisesRegex`)

big-c-note replied:
Ok great

jakevdp commented on tests/scipy_stats_test.py:
```diff
@@ -889,5 +891,110 @@ def evaluate_kde(kde, x):
+    else:
+      rng = jtu.rand_default(self.rng())
+    # Rounding to create common values more probable
+    args_maker = lambda: [jnp.round(rng(x_shape, x_dtype)), axis, nan_policy]
```
`axis` and `nan_policy` are static arguments, so we should leave them out of `args_maker` and instead pass them to the functions via closure or `partial`. Then the same functions & `args_maker` can be passed to both `_CheckAgainstNumpy` and `_CompileAndCheck`.

big-c-note replied:
Yes, I see this now thanks.

jakevdp reviewed (changes requested):
Thanks - this is a great start! I think the core of the approach here is quite solid. I have a number of comments below, some very minor and some with more substance. Please feel free to ask if anything is unclear. Thanks!

yashk2810 committed:
Only run the pjit gda mesh check when jax_enable_checks is enabled and cache _create_mesh_pspec_sharding

PiperOrigin-RevId: 460775764

jakevdp committed:
Populate long_description for jax & jaxlib

shoyer committed:
Fix stacklevel for _check_user_dtype_supported

Before this change, JAX generates warnings referring to the _internal_
line inside JAX where `_check_user_dtype_supported` is called:

    In [1]: import jax.numpy as jnp

    In [2]: jnp.zeros(2, jnp.float64)
    /Users/shoyer/dev/jax/jax/_src/numpy/lax_numpy.py:1976: UserWarning: Explicitly requested dtype <class 'jax.numpy.float64'> requested in zeros is not available, and will be truncated to dtype float32. To enable more dtypes, set the jax_enable_x64 configuration option or the JAX_ENABLE_X64 shell environment variable. See https://github.com/google/jax#current-gotchas for more.
    lax_internal._check_user_dtype_supported(dtype, "zeros")
    Out[2]: DeviceArray([0., 0.], dtype=float32)

Now it generates warnings referring to user code:

    In [1]: import jax.numpy as jnp

    In [2]: jnp.zeros(2, jnp.float64)
    <ipython-input-2-dc38008f18e8>:1: UserWarning: Explicitly requested dtype <class 'jax.numpy.float64'> requested in zeros is not available, and will be truncated to dtype float32. To enable more dtypes, set the jax_enable_x64 configuration option or the JAX_ENABLE_X64 shell environment variable. See https://github.com/google/jax#current-gotchas for more.
    jnp.zeros(2, jnp.float64)
    Out[2]: DeviceArray([0., 0.], dtype=float32)

big-c-note commented on tests/scipy_stats_test.py:
```diff
@@ -889,5 +891,110 @@ def evaluate_kde(kde, x):
+      The `scipy.stats.mode` function returns a tuple:
+       `namedtuple("ModeResult", ("mode", "count"))`
+      So, we need to check both parts.
+      """
```
Ah yes, I see how this works now, thanks

big-c-note committed:
reorders imports, fixes typing, ensures DeviceArray, simplifies logic, writes error messages

big-c-note commented:
Thanks for the detailed review and your time @jakevdp . I'm enjoying the project.

Responded to your points and implemented.

No worries on the cla.

If you've got any recommendations on issues to work on next, I'd like to either get (a tad) closer to the compiler or implement something that makes use of `grad` within the code base.

Let me know if any further updates would be nice here

big-c-note committed:
removing redundant logic

big-c-note commented on jax/_src/scipy/stats/_core.py:
```diff
@@ -66,8 +66,6 @@ def _mode_helper(x: jnp.ndarray) -> Tuple[jnp.ndarray, jnp.ndarray]:
     axis = 0
   x_shape = list(np.shape(x))
-  num_dims = len(x_shape)
-  axis = canonicalize_axis(axis, num_dims)
```
The logic is accounted for by `moveaxis` I believe

marcvanzee committed:
[jax2tf] Do not use tf.math.invert_permutation when enable_xla=False.
TFjs currently does not provide support for InvertPermutation: https://github.com/tensorflow/tfjs/issues/6395.

PiperOrigin-RevId: 460914779

jakevdp commented on jax/_src/scipy/stats/_core.py:
```diff
@@ -0,0 +1,72 @@
+      x,
+      "In order to best JIT compile `mode`, we cannot know whether `x` contains nans. "
+      "Please check if nans exist in `x` outside of the `mode` function."
+    )
```
Since the function is wrapped in `jit` by default, this will always raise. Better would be either (1) raise `NotImplementedError` directly rather than using `concrete_or_error`, or (2) not wrap `mode` in `jit` so that `mode = 'raise'` can be implemented for concrete inputs `x`. I'm fine either way.

jakevdp commented on jax/_src/scipy/stats/_core.py:
```diff
@@ -0,0 +1,72 @@
+    )
+
+  if x.size == 0:
+    return ModeResult(jnp.array([]), jnp.array([]))
```
We should ensure that the dtype of the arrays here match the expected result.

Also, not all empty arrays are the same shape – e.g. the output of mode for an array of shape `[0, 5]` along `axis=1` should probably have shape `[0, 1]`

big-c-note replied:
> Also, not all empty arrays are the same shape – e.g. the output of mode for an array of shape [0, 5] along axis=1 should probably have shape [0, 1]

Looking [here](https://github.com/scipy/scipy/blob/4cf21e753cf937d1c6c2d2a0e372fbc1dbbeea81/scipy/stats/_stats_py.py#L411-L412) in `scipy.stats.mode` for loose inspiration into the logic that produces the shape for an empty array.

Check this below, and see if this is what you mean. This is our current behavior - which produces the same result as `scipy.stats.mode`.
```
import numpy as np
from scipy.stats import mode as scipy_mode

from jax._src.scipy.stats._core import mode as jax_mode


cat = np.array([[], []])
cat = cat.reshape((0,2))
cat.shape
#  (0, 2)

scipy_mode(cat, 1)[0].shape
#  (0,)

jax_mode(cat, 1)[0].shape
#  (0,)
```
> We should ensure that the dtype of the arrays here match the expected result.

Did you mean explicitly within the code via an assert statement? If so, this is the only thing I didn't implement in what I'm about to push. I verified here:
```
cat = np.array([1], dtype=np.float16)
scipy_fun(cat, axis=0)[0].dtype
#  dtype('float16')
jax_fun(cat, axis=0)[0].dtype
#  dtype('float16')
```

jakevdp commented on jax/_src/scipy/stats/_core.py:
```diff
@@ -0,0 +1,72 @@
+    x = lax.reshape(x, (np.size(x),))
+    axis = 0
+  x_shape = list(np.shape(x))
+  x_shape[axis] = 1
```
We should validate the axis before using it (`canonicalize_axis` is a useful utility that raises informative errors in common cases like non-static inputs)

big-c-note replied:
Yes, I see this now, thank you

jakevdp commented on tests/scipy_stats_test.py:
```diff
@@ -889,5 +890,84 @@ def evaluate_kde(kde, x):
+    jtu.cases_from_list(
+      {
+        "testcase_name":
+          f"""{x_shape};{x_dtype};{axis};{nan_policy};{contains_nans} """,
```
A couple things here:
- We generally use underscores rather than semicolons in test names.
- please use the `format_shape_dtype_string` utility for the shape and type
- please add parameter names to the test string; e.g. `_axis={axis}_`
- Use single quotes for a one-line string

big-c-note replied:
Word, thanks

jakevdp reviewed (changes requested):
Looking good! A few more comments

skye committed:
Update WORKSPACE and setup.py in preparation for 0.3.15 jax/jaxlib release

jakevdp committed:
CI: use minimum jaxlib in upstream-ci build

jakevdp committed:
lax_numpy_test: make compatible with numpy 1.24-dev

big-c-note committed:
simplifies test, raises NotImplemented error for nan_pilicy raise

big-c-note commented on jax/_src/scipy/stats/_core.py:
```diff
@@ -0,0 +1,72 @@
+    return ModeResult(jnp.array([]), jnp.array([]))
+  if axis is None or _isscalar(x):
+    x = lax.reshape(x, (np.size(x),))
+    axis = 0
```
Ah, ok nice catch. I see the  usefulness of `canonicalize_axis` now

gnecula committed:
[dynamic-shapes] Split dynamic_api_test.py

PiperOrigin-RevId: 461109288

tlu7 committed:
[sparse] Update `_validate_coo_mhlo` in gpu_sparse.

PiperOrigin-RevId: 461111317

big-c-note committed:
checks axis

big-c-note commented:
Shoot, that cla error does not belong to me - was trying to get all the latest changes in the branch, and I brought in a commit from a-googler.

Anyway, it's close, but I did notice two oddities.
1) `jax.scipy.stats.mode(2, axis=0)` produces `ValueError` for axis out of bounds. `scipy.stats.mode(2, axis=0)` produces a result.

I'm going to hack on that this weekend and check through a few other edge cases.

2) Am I using `_wraps` for docstrings appropriately? In an ipython debug session I was running `scipy.stats.mode??` and seeing documentation, but `jax.scipy.stats.mode??` gave me back code. Maybe just strageness with ipython but throwing it out there.

jakevdp commented:
> Am I using _wraps for docstrings appropriately? In an ipython debug session I was running scipy.stats.mode?? and seeing documentation, but jax.scipy.stats.mode?? gave me back code. Maybe just strageness with ipython but throwing it out there.

double question marks in IPython shows source code, while single question marks show the docstring. It sounds like this is working as expected, given that in scipy the docstring literal is in the source code, while in JAX it's included programatically.

jakevdp commented:
Any update? You should be able to get your branch in order using something like this
```bash
$ git checkout main
$ git pull upstream main # <-- assuming upstream is the name you used to point to the main jax repo
$ git checkout adds_scipy_stats_mode
$ git rebase main
```

jakevdp commented:
Thanks for the work on this – closing in favor of https://github.com/google/jax/pull/12683
